cwltool/cuda.py

```diff
@@ -8,23 +8,48 @@
 from .utils import CWLObjectType
 
 
+def cuda_device_count() -> str:
+    """Determine the number of attached CUDA GPUs."""
+    # For the number of GPUs, we can use the following query
+    cmd = ["nvidia-smi", "--query-gpu=count", "--format=csv,noheader"]
+    try:
+        # This is equivalent to subprocess.check_output, but use
+        # subprocess.run so we can use separate MagicMocks in test_cuda.py
+        proc = subprocess.run(cmd, stdout=subprocess.PIPE, check=True)  # nosec
+    except Exception as e:
+        _logger.warning("Error checking number of GPUs with nvidia-smi: %s", e)
+        return "0"
+    # NOTE: On a machine with N GPUs the query return N lines, each containing N.
+    return proc.stdout.decode("utf-8").split("\n")[0]
+
+
 def cuda_version_and_device_count() -> Tuple[str, int]:
     """Determine the CUDA version and number of attached CUDA GPUs."""
+    count = int(cuda_device_count())
+
+    # Since there is no specific query for the cuda version, we have to use
+    # `nvidia-smi -q -x`
+    # However, apparently nvidia-smi is not safe to call concurrently.
+    # With --parallel, sometimes the returned XML will contain
+    # <process_name>\xff...\xff</process_name>
+    # (or other arbitrary bytes) and xml.dom.minidom.parseString will raise
+    # "xml.parsers.expat.ExpatError: not well-formed (invalid token)"
+    # So we either need to use `grep -v process_name` to blacklist that tag,
+    # (and hope that no other tags cause problems in the future)
+    # or better yet use `grep cuda_version` to only grab the tags we will use.
+    cmd = "nvidia-smi -q -x | grep cuda_version"
     try:
-        out = subprocess.check_output(["nvidia-smi", "-q", "-x"])  # nosec
+        out = subprocess.check_output(cmd, shell=True)  # nosec
```

Comment:
I would prefer that instead of using a subshell it captures the output and does the substring search in Python.  For a simple substring search this really is as simple as something like:
```
out = subprocess.check_output(["nvidia-smi", "-q", "-x"])
if "cuda_version" not in out:
   raise ...
```

Reply:
Actually instead of using minidom, to pull a single tag it would be easier to just use a regular expression to look for `<cuda_version>`.